File: HPWorkpath_v1.6.3/Samples/ExampleExtensions/source/GoogleSigninSample/src/com/hp/workpath/sample/googlesigninsample/task/RequestTokenTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.googlesigninsample.task;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.hp.workpath.sample.googlesigninsample.MainActivity;
import com.hp.workpath.sample.googlesigninsample.model.AccountInfo;
import com.hp.workpath.sample.googlesigninsample.model.ClientInfo;

import java.io.IOException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class RequestTokenTask {

    private static final String TAG = MainActivity.TAG;

    private ClientInfo mClientInfo;
    private String mCode;
    private TaskInterface mTaskInterface;
    private Throwable mThrowable;
    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public RequestTokenTask(@NonNull ClientInfo clientInfo, @NonNull String code, @NonNull TaskInterface taskInterface) {
        this.mClientInfo = clientInfo;
        this.mCode = code;
        this.mTaskInterface = taskInterface;
    }

    public void taskExecute() {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    try {
                        OkHttpClient client = new OkHttpClient().newBuilder()
                                .connectTimeout(60, TimeUnit.SECONDS)
                                .readTimeout(60, TimeUnit.SECONDS)
                                .writeTimeout(60, TimeUnit.SECONDS)
                                .build();

                        RequestBody requestBody = new MultipartBody.Builder()
                                .setType(MultipartBody.FORM)
                                .addFormDataPart("code", mCode)
                                .addFormDataPart("client_id", mClientInfo.getClientId())
                                .addFormDataPart("redirect_uri", mClientInfo.getRedirectUri())
                                .addFormDataPart("grant_type", mClientInfo.getGrantType())
                                .addFormDataPart("code_verifier", mClientInfo.getCodeVerifier())
                                .addFormDataPart("client_secret", mClientInfo.getClientSecret())
                                .build();

                        Request request = new Request.Builder()
                                .url(mClientInfo.getTokenUri())
                                .addHeader("Content-Type", "application/x-www-form-urlencoded")
                                .method("POST", RequestBody.create(null, new byte[0]))
                                .post(requestBody)
                                .build();


                        Response response = client.newCall(request).execute();
                        if (response != null) {
                            if (response.isSuccessful()) {
                                String responseBody = response.body().string();
                                Gson gson = new Gson();
                                onPostExecute(gson.fromJson(responseBody, AccountInfo.class));
                            } else {
                                mThrowable = new Exception(response.body().string());
                                onPostExecute(null);
                            }
                        } else {
                            mThrowable = new Exception("Request token response is null");
                            onPostExecute(null);
                        }
                    } catch (Throwable t) {
                        mThrowable = t;
                        onPostExecute(null);
                        executor.shutdown();
                    }
                }
            });
        } catch (Throwable t) {
            mThrowable = t;
            onPostExecute(null);
            executor.shutdown();
        }
    }

    protected void onPostExecute(AccountInfo accountInfo) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (accountInfo == null) {
                    mTaskInterface.onFailure(mThrowable);
                } else {
                    mTaskInterface.receivedToken(accountInfo);
                }
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleExtensions/source/GoogleSigninSample/src/com/hp/workpath/sample/googlesigninsample/task/RequestTokenTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.googlesigninsample.task

import com.google.gson.Gson
import com.hp.workpath.sample.googlesigninsample.MainActivity
import com.hp.workpath.sample.googlesigninsample.model.AccountInfo
import com.hp.workpath.sample.googlesigninsample.model.ClientInfo
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

class RequestTokenTask(private val mClientInfo: ClientInfo, private val mCode: String, private val mTaskInterface: TaskInterface) {
    private var mThrowable: Throwable? = null

    suspend fun execute() {
        var accountInfo: AccountInfo? = null
        val client = OkHttpClient().newBuilder()
                .connectTimeout(60, TimeUnit.SECONDS)
                .readTimeout(60, TimeUnit.SECONDS)
                .writeTimeout(60, TimeUnit.SECONDS)
                .build()
        val requestBody: RequestBody = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("code", mCode)
                .addFormDataPart("client_id", mClientInfo.clientId)
                .addFormDataPart("redirect_uri", mClientInfo.redirectUri)
                .addFormDataPart("grant_type", mClientInfo.grantType)
                .addFormDataPart("code_verifier", mClientInfo.codeVerifier)
                .addFormDataPart("client_secret", mClientInfo.clientSecret)
                .build()
        val request = Request.Builder()
                .url(mClientInfo.tokenUri)
                .addHeader("Content-Type", "application/x-www-form-urlencoded")
                .method("POST", RequestBody.create(null, ByteArray(0)))
                .post(requestBody)
                .build()
        try {
            val response = client.newCall(request).execute()
            if (response != null) {
                if (response.isSuccessful) {
                    val responseBody = response.body()!!.string()
                    val gson = Gson()
                    accountInfo = gson.fromJson(responseBody, AccountInfo::class.java)
                } else {
                    Throwable(response.body()?.string())
                }
            } else {
                Throwable("Request token response is null")
            }
        } catch (e: IOException) {
            mThrowable = e
        }
        onPostExecute(accountInfo)
    }

    private suspend fun onPostExecute(accountInfo: AccountInfo?) {
        if (accountInfo == null) {
            mTaskInterface.onFailure(mThrowable)
        } else {
            mTaskInterface.receivedToken(accountInfo)
        }
    }

    companion object {
        private val TAG: String = MainActivity.TAG
    }
}